### PR TITLE
[SYCL] Diagnose uses of zero length arrays

### PR DESCRIPTION
**Pull request by Fznamznon**
Adds diagnosing on attempt to use zero length arrays, pointers, refs, arrays
of them and structs/classes containing all of it.
A couple of assertions were removed from SemaSYCL because the new
diagnostics are deferred and emitted later than SemaSYCL routines.
In case a struct/class with zero length array is used this emits a set
of notes pointing out how zero length array got into used struct, like
this:
```
struct ContainsArr {
  int A[0]; // note: field of illegal type declared here
};
struct Wrapper {
  ContainsArr F; // note: within field of type ContainsArr declared here
  // ...
}

// Device code
Wrapper W;
W.use(); // error: zero-length arrays are not permitted

```
Total deep check of each used declaration may result in double
diagnosing at the same location.

### CODE REVIEW

File: clang/lib/Sema/SemaSYCL.cpp [hunk 1]
```diff
@@ -422,12 +422,99 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+                             diag::note_illegal_type_decl_here)
+            << D << D->getType();
+      }
+    }
+
+    return ErrorFound;
+  };
+
+  // In case we have a Record used do the DFS for a bad field
+  SmallVector<const ValueDecl *, 4> StackForRecursion;
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:463]: We shouldn't be doing this level of deep checking in the CFE.  Walking the AST this deep is exactly the stuff @AaronBallman  and I have been warning against/trying to stop us from doing.

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:463]: Hmm... I guess at second look, this is going through the fields instead of calls, so it isn't as bad as I thought.

I'd still recommend doing this differently, I'd suggest diagnosing by calling RecordDecl::CheckSYCLTypes (or something, the point is that it loops through its fields and checks, plus recursing down CheckSYCLTypes when necessary), then having it return whether it diagnosed or not so that you can emit the note if you care to.

Printing EVERY error, or the history for every error is likely overly verbose anyway.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > Hmm... I guess at second look, this is going through the fields instead of calls, so it isn't as bad as I thought.

Yes, that is going through fields. Actually I didn't invent that, it seems clang already does such thing for OpenCL kernel arguments https://github.com/intel/llvm/blob/35ccdd5a6b3c467cd39bfb14f0abc7ef2bc20667/clang/lib/Sema/SemaDecl.cpp#L8958 .

> I'd suggest diagnosing by calling RecordDecl::CheckSYCLTypes (or something, the point is that it loops through its fields and checks, plus recursing down CheckSYCLTypes when necessary), then having it return whether it diagnosed or not so that you can emit the note if you care to.

Do you mean we should stop recursion once the first field of illegal type is found?

Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:463]: Ah, weird!  I'm surprised that made it by community review, though I do know that the OpenCL code owners are a bit lax as far as code review effectiveness.  That said, if instead of this patch we could just generalize the OpenCL version, at least we wouldn't be doing further harm (AND this would be a good opportunity to upstream this immediately instead of doing here).

I would probably give up on the notes after the 1st in each struct/path.  Blitting out the 'how we got here' non stop is just going to be a giant mess of notes I suspect.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > That said, if instead of this patch we could just generalize the OpenCL version

I will try that, but I'm pretty sure this will be pretty annoying to do (if even possible!), since OpenCL code seems very... OpenCL specific.

> we wouldn't be doing further harm

If you think this does harm, I'm open for any ideas. Do the recursive search for bad struct fields is the best way to solve the problem that I could figure out.

> this would be a good opportunity to upstream this immediately instead of doing here

I believe the best way to upstream this is to ask developers of other single source offload models (CUDA/OMP/maybeHIP?) if they need such thing and generalize the function I've added across them. OpenCL folks already have their diagnostic, the generalization will likely complicate the code and I think they don't care about SYCL and therefore the change won't bring anything good for them, so this will end up rejected or it will take forever to merge.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: Yeah, the first problem with generalization is that OpenCL code emits errors for pointers met inside struct, whereas SYCL needs to take pointee type and go further instead.

File: clang/lib/Sema/Sema.cpp [hunk 2]
```diff
@@ -1931,6 +1931,17 @@ void Sema::checkTypeSupport(QualType Ty, SourceLocation Loc, ValueDecl *D) {
   if (isUnevaluatedContext() || Ty.isNull())
     return;
 
+  // The original idea behind checkTypeSupport function is that unused
+  // declarations can be replaced with an array of bytes of the same size during
+  // codegen, however this is not yet implemented for SYCL and cause production
+  // of incorrect SPIR-V.
+  // In addition such replacement doesn't seem to be possible for types without
+  // constant byte size like zero length arrays. So, do a deep check for SYCL.
```
Comment (AaronBallman) [inline clang/lib/Sema/Sema.cpp:1939]: What about flexible array members?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:463]: Ultimately, where we want this diagnostic to live is within SemaDecl.cpp when processing the declaration of a zero-length array. The fact that we have to wait until the entire TU is complete before deciding whether we can or cannot diagnose a declaration elsewhere is a design flaw.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > Ultimately, where we want this diagnostic to live is within SemaDecl.cpp when processing the declaration of a zero-length array. The fact that we have to wait until the entire TU is complete before deciding whether we can or cannot diagnose a declaration elsewhere is a design flaw.

With current SYCL spec design we cannot do that. We don't know what is device code and what is not until `parallel_for` template is ready. This patch at least doesn't extend SemaSYCL kernels processing and uses deferred diagnostics like other single source offloading models.

File: clang/lib/Sema/Sema.cpp (hunk 2 above)
Comment (Fznamznon) [reply clang/lib/Sema/Sema.cpp:1939]: Can be diagnosed this way as well. I thought about that, but I left it for next patches to keep this one simpler/smaller.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > With current SYCL spec design we cannot do that. We don't know what is device code and what is not until parallel_for template is ready. 

Yup, and we were asked explicitly to push back when the SYCL spec is designed in a way that's not reasonable to our implementation. CC @pvchupin for his awareness.

> This patch at least doesn't extend SemaSYCL kernels processing and uses deferred diagnostics like other single source offloading models.

I'm not blocking this review so there may not be anything actionable here for you, but the reason I'm pushing back is because this seems to be building on top of something we already know is not likely to be acceptable for upstreaming. So you're right that it doesn't make new problems, but it does make it harder for us to go back and fix that foundation. :-(

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > we already know is not likely to be acceptable for upstreaming

Why so? The whole `checkTypeSupport` routine that I'm extending is in upstream clang, and it uses the approach "emit a deferred diagnostic for unsupported type", it works not only for SYCL, but for OMP as well. The only new thing I brought here is recursive check for struct fields (using the idea that already used in community code). The reason why SYCL prohibits zero length arrays is that there is no memory allocation on GPUs (and any other weird devices where you can run opencl/sycl).

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > Why so?

As mentioned; the compiler should not have to process the entire TU before deciding whether to diagnose the declaration of some earlier construct or not. That's a deficiency of the language extension design; C++ is not a multi-pass programming language.

You piqued my interest though as to how this got into community in the first place... https://reviews.llvm.org/D109315 renamed to `checkTypeSupport()` but https://reviews.llvm.org/D74387 is what introduced the notion originally. I see John McCall was on that first review, but despite having a considerable number of comments, he never signed off on it before it landed. I didn't check all of his comments to see if this was discussed or not.

All that said, I think the *desire* for the diagnostic is reasonable and something I'm in support of. I'm pushing back on the SYCL making it impossible to implement reasonable diagnostics without having to make multiple passes over an entire TU.

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > C++ is not a multi-pass programming language.

But maybe SYCL is?

I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in `checkTypeSupport()`. I agree that SYCL could define device code entry more precisely like CUDA does with its "device/host/etc" attributes instead of "everything called from template function with this name magically becomes device code". But even CUDA has places where deferred diagnostics needed, that means even there it is not possible to understand where is device code until everything is parsed.

Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:463]: > But maybe SYCL is?

SYCL is supposed to be an extension to C++ and I'd question whether redefining the compilation model is a conforming extension in practice.

> I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in checkTypeSupport(). 

I agree! The difference is: deferred diagnostics typically do not require deferring to the end of the TU, but instead to the end of the function definition or class definition (or perhaps instantiation thereof). Function bodies and class bodies are generally far more bounded than an entire TU. (Consider how well this scales to TUs that are code generated; you could get 1 million lines of code that forms an AST we have to walk multiple times only to discover that there was little or no SYCL code anywhere in the TU to begin with.)

Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:469]: > SYCL is supposed to be an extension to C++

Okay, I didn't see SYCL from this point of view and rather perceived it as a separate C++ based language. 

@erichkeane , I've reduced number of emitted notes, all notes are emitted only when the first error is discovered. Please let me know if this became better.

**Comment by Fznamznon** [comment]
I'm also thinking about adding a new error message that says "zero-length arrays are not permitted in *SYCL*" because right now it says "zero-length arrays are not permitted in C++" which might be confusing because the host code is also C++.
I know that I probably was the one who suggested to re-use the community diagnostic, because re-using is always better for upstream, but still....
Reviewers, please let me know your thoughts.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (erichkeane) [reply clang/lib/Sema/SemaSYCL.cpp:469]: I'm ok with it, I'll leave the decision on the rest to the code owners.

**Comment by AaronBallman** [comment]
> I'm also thinking about adding a new error message that says "zero-length arrays are not permitted in _SYCL_" because right now it says "zero-length arrays are not permitted in C++" which might be confusing because the host code is also C++. I know that I probably was the one who suggested to re-use the community diagnostic in similar check inside SemaSYCL, because re-using is always better for upstream, but still.... Reviewers, please let me know your thoughts.

I think the change in diagnostic text isn't a bad idea. Zero-length arrays are an extension in C++ that Clang supports by default, so the current wording could be confusing.

**Comment by Fznamznon** [comment]
> I think the change in diagnostic text isn't a bad idea. Zero-length arrays are an extension in C++ that Clang supports by default, so the current wording could be confusing.

Done then.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 1 above)
Comment (keryell) [reply clang/lib/Sema/SemaSYCL.cpp:469]: > > C++ is not a multi-pass programming language. [@AaronBallman]
> 
> But maybe SYCL is? [@Fznamznon]

Not really. This is an implementation detail.
 
> I got your point. I was just only trying to say that need to parse the entire TU to understand what is what is not only SYCL problem. Deferred diagnostics are widely used across clang, not only in `checkTypeSupport()`. I agree that SYCL could define device code entry more precisely like CUDA does with its "device/host/etc" attributes instead of "everything called from template function with this name magically becomes device code". 

Argh! I have a stroke! I need some oxygen! :-) I think you are ready to send your résumé to NVIDIA at that point. @brycelelbach you have a candidate! ;-)
Single-source and plain C++ without extension is the essence of SYCL.

Even handling templates in plain C++ requires some evaluation delay to provide better diagnostics.

> But even CUDA has places where deferred diagnostics needed, that means even there it is not possible to understand where is device code until everything is parsed.

And I guess that if NVIDIA is up-streaming in the future their pure NVC++ they will have a  somehow similar issue on this aspect. Any feedback from NVIDIA compiler team on this?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 3]
```diff
@@ -10129,6 +10129,8 @@ def note_within_field_of_type : Note<
   "within field of type %0 declared here">;
 def note_illegal_field_declared_here : Note<
   "field of illegal %select{type|pointer type}0 %1 declared here">;
+def note_illegal_type_decl_here : Note<
+  "declaration %0 of illegal type %1 is here">;
```
Comment (AaronBallman) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:10133]:
```suggestion
  "declaration %0 of invalid type %1 is here">;
```
Let's not block on this if it turns out to be a contentious suggestion. We usually try to avoid using "illegal" in diagnostics because that has implications in different parts of the world, but we don't always succeed (as the preceding note shows).

Actually, I don't think we need this new note at all. We can dodge it entirely by using `note_declared_at` from `DiagnosticCommonKinds.td`.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 4]
```diff
@@ -422,12 +422,111 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+                               diag::note_illegal_type_decl_here)
+              << D << D->getType();
+        }
+      }
+    }
+
+    return ErrorFound;
+  };
+
+  // In case we have a Record used do the DFS for a bad field
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:468]:
```suggestion
  // In case we have a Record used do the DFS for a bad field.
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 5]
```diff
@@ -422,12 +422,111 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+          EmitHistory();
+        NeedToEmitNotes = false;
+      }
+    }
+
+    if (const auto *RecDecl = NextTy->getAsRecordDecl()) {
+      if (auto *NextFD = dyn_cast<FieldDecl>(Next))
+        History.push_back(NextFD);
+      // When nullptr is discovered, this means we've gone back up a level, so
+      // the history should be cleaned
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:522]:
```suggestion
      // the history should be cleaned.
```

File: clang/lib/Sema/SemaSYCL.cpp [hunk 6]
```diff
@@ -422,12 +422,111 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+      StackForRecursion.push_back(nullptr);
+      for (const auto *FD : RecDecl->fields())
+        StackForRecursion.push_back(FD);
```
Comment (AaronBallman) [inline clang/lib/Sema/SemaSYCL.cpp:525]: llvm::copy?

**Review by AaronBallman — COMMENTED** [review]
Just nits from me.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10133]: > We usually try to avoid using "illegal" in diagnostics because that has implications in different parts of the world

Thank you for noting this. I didn't know about that.

> Actually, I don't think we need this new note at all. 

I know it is probably not essential, but I wanted to show a concrete type and declaration's to the user in the message, I think it looks much clearer, especially when declaration of unsupported type depends on some template parameters.
If you still think that it is better to use `note_declared_at` I ll switch to it.

Comment (AaronBallman) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td:10133]: We try to reuse existing diagnostics as much as possible because we can (and have, repeatedly) run out of diagnostic IDs, so my default as a reviewer is "how can we use an existing diagnostic instead of adding a new one". While I think your wording is more precise than "declared here" and looks better in isolation, notes are associated with a specific warning or error diagnostic already. So, to me, it really boils down to whether we think the %0 and %1 are necessary. Based on the test cases, I don't think they are because they all look like this:
```
type decl; // declaration 'decl' of illegal type 'type' is here
```
where the diagnostic is repeating what's already on the line of code. Given that I don't think declaring zero-sized arrays is a common thing, and that it's not likely you'll get multiple declarations on the same line, I think the existing note is sufficiently clear and we should go with it. WDYT?

(Btw, the first time somebody complains about the note being confusing, I'm fully on board with your proposed wording.)

File: clang/lib/Sema/SemaSYCL.cpp (hunk 6 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done

File: clang/lib/Sema/SemaSYCL.cpp (hunk 5 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done

File: clang/lib/Sema/SemaSYCL.cpp (hunk 4 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp]: Done.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td (hunk 3 above)
Comment (Fznamznon) [reply clang/include/clang/Basic/DiagnosticSemaKinds.td]: And, done.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 7]
```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+  if (const auto *CAT = SemaRef.getASTContext().getAsConstantArrayType(Ty))
+    return CAT->getSize() == 0;
   return false;
 }
 
+// NOTE: In case this function gets upstreamed and other single source
+// offloading models also need it, it can be generalized via using of
+// `targetDiag` instead of `SYCLDiagIfDeviceCode` and removing assertion at the
+// beginning.
+void Sema::deepTypeCheckForSYCLDevice(SourceLocation UsedAt,
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:435]: This is a nit but I think renaming this to `CheckArrayType `or something along that is better . Or are you naming it in a generic manner to future proof it in case you need this check for other types?

File: clang/lib/Sema/SemaSYCL.cpp [hunk 8]
```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
   return false;
 }
 
-static bool isZeroSizedArray(QualType Ty) {
-  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
-    return CATy->getSize() == 0;
+static bool isZeroSizedArray(Sema &SemaRef, QualType Ty) {
```
Comment (elizabethandrews) [inline clang/lib/Sema/SemaSYCL.cpp:425]: Can you please explain why this change was required?

File: clang/test/SemaSYCL/zero-length-arrays.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+  });
+  WrapperOfWrapper St;
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<SimpleStruct, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 2{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class SimpleStruct>([=]{
+      // expected-note@+1 {{within field of type 'WrapperOfWrapper' declared here}}
+      (void)St.F.BadArray; // expected-error 4{{zero-length arrays are not permitted in SYCL device code}}
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/zero-length-arrays.cpp:101]: Why is this diagnostic generated 4 times for this declaration?

File: clang/test/SemaSYCL/zero-length-arrays.cpp [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  Templated<1 - 1, double> Weirdo;
+  Templated<0, float> Zero;
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<UseTemplated, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 4{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+  // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class UseTemplated>([=]{
+      (void)OK; // No errors expected
+      (void)Zero; // expected-error 2{{zero-length arrays are not permitted in SYCL device code}}
```
Comment (elizabethandrews) [inline clang/test/SemaSYCL/zero-length-arrays.cpp:124]: Same here. Why 2 diagnostics for the same thing? Is the additional diagnostic for instantiated template?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 7 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:435]: > Or are you naming it in a generic manner to future proof it in case you need this check for other types?

Yeah.

File: clang/test/SemaSYCL/zero-length-arrays.cpp (hunk 10 above)
Comment (Fznamznon) [reply clang/test/SemaSYCL/zero-length-arrays.cpp:124]: Because there are two zero-length arrays inside this object. An error is generated for each of them (notes are not emitted after first error for the same object).

File: clang/test/SemaSYCL/zero-length-arrays.cpp (hunk 9 above)
Comment (Fznamznon) [reply clang/test/SemaSYCL/zero-length-arrays.cpp:101]: Two errors for `WrapperOfWrapper` struct because it contains two arrays, one for use of field `F` because it has an array inside, another one for use of `BadArray` because it is array.

Comment (elizabethandrews) [reply clang/test/SemaSYCL/zero-length-arrays.cpp:101]: Oh right. Thanks for explaining!

File: clang/lib/Sema/SemaSYCL.cpp (hunk 8 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:425]: I don't quite remember all the details, so my explanation may not be quite correct, but what I remember is that we shouldn't do `dyn_cast` on types because in AST just say type A and typedef-ed/qualified/sugared type A are not same types. The right way to do it like we did in the past - https://github.com/intel/llvm/pull/2082/files .

Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:425]: Oh yea. I remember now - https://github.com/intel/llvm/pull/2072#discussion_r452201923

File: clang/lib/Sema/SemaSYCL.cpp [hunk 11]
```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
+      }
+    };
+    if (Check(NextTy, Next)) {
+      if (NeedToEmitNotes)
+        EmitHistory();
+      NeedToEmitNotes = false;
+    }
+
+    // In case pointer/array/reference type is met get pointeetype, then proceed
+    // with that type.
```
Comment (smanna12) [inline clang/lib/Sema/SemaSYCL.cpp:502]: Can you update the comment here? It is  not clear.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 12]
```diff
@@ -422,12 +422,108 @@ bool Sema::isKnownGoodSYCLDecl(const Decl *D) {
...
-static bool isZeroSizedArray(QualType Ty) {
-  if (const auto *CATy = dyn_cast<ConstantArrayType>(Ty))
-    return CATy->getSize() == 0;
+static bool isZeroSizedArray(Sema &SemaRef, QualType Ty) {
+  if (const auto *CAT = SemaRef.getASTContext().getAsConstantArrayType(Ty))
+    return CAT->getSize() == 0;
   return false;
 }
 
+// NOTE: In case this function gets upstreamed and other single source
```
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:431]: I think upstreaming of non-trivial SYCL-related patch is a very long shot that probably may not succeed anyway while our customers will face non-user friendly crashes/problems later in the toolchain, so this should be done as a separate task.
In addition, after Aaron's comments - https://github.com/intel/llvm/pull/4888#discussion_r744929969 , I don't think it is worth it at all.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 11 above)
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:502]: Could you suggest better wording, please?

File: clang/lib/Sema/SemaSYCL.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:431]: I'm not sure what the required protocol here is. Unless we have have a non-negotiable strict deadline, I believe we should be upstreaming whatever can be upstreamed (even if it delays patch). If the patch in its current form is not up-streamable, that should be fixed. From what I can see we already have required infrastructure for this patch upstream (like `checkTypeSupport`), so I don't see what is blocking an attempt to upstream it. 

@AaronBallman I am not sure I understand your comment here - https://github.com/intel/llvm/pull/4888#discussion_r744929969. This patch is building on top of something which has already been upstreamed. Are you suggesting that will be reverted in the future?

[the remark this answers is not in the document]
Comment (elizabethandrews) [reply clang/lib/Sema/SemaSYCL.cpp:431]: Tagging @bader and @zahiraam for their input as well since they drive upstreaming efforts

File: clang/lib/Sema/SemaSYCL.cpp (hunk 11 above)
Comment (smanna12) [reply clang/lib/Sema/SemaSYCL.cpp:502]: I looked at comment again. I misinterpreted. Sorry for the confusion. Do not need to update it. Thanks.

File: clang/lib/Sema/SemaSYCL.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:431]: > @AaronBallman I am not sure I understand your comment here - #4888 (comment). This patch is building on top of something which has already been upstreamed. Are you suggesting that will be reverted in the future?

That comment was to suggest that I'm not certain the original code should have landed upstream because it has design issues, and building on top of a questionable foundation is something we want to avoid. However, because it's already landed (for a reasonable amount of time) and downstreams are building on top of it, I think it would cause undue churn to just go and revert it. (That doesn't mean it won't get reverted in the future if someone upstream sees the code and decides it's a problem they won't live with.)

Our goal is to try to upstream everything we can and upstreaming this change fits that goal. To me, the only danger with upstreaming this patch would be that someone in community takes issue with the foundational bits and asks you to rewrite those (which would be a good technical outcome but is not likely planned or particularly trivial effort).

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:431]: So, did I get it right that I should try to upstream this patch?

[the remark this answers is not in the document]
Comment (AaronBallman) [reply clang/lib/Sema/SemaSYCL.cpp:431]: I'm curious what @bader and @zahiraam think, but I would say it's reasonable to upstream the patch (but it could lead to unplanned work).

[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:431]: I think whether this patch is up-streamed now or later there will always be un-planed work. It would be my preference to wait for it to be up-streamed only because it was a request from community to do a couple of things before SemaSYCL.cpp gets into an acceptable shape. The handling for special classes (this has already been done) but also the mix if "sema" and "codegen" work that's happening in SemaSYCL.cpp.

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:431]: Just to make sure we are on the same page - this patch actually doesn't depend on things already implemented in SemaSYCL (like kernel generation). I could put the function that this patch adds to SemaSYCL anywhere else, I just did that because it is called only for SYCL. So nothing blocks from upstreaming this patch in parallel with kernel generation.

[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp:431]: Then upstreaming it is fine. Thanks.

[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/lib/Sema/SemaSYCL.cpp:431]: Okay then. I've moved it to https://reviews.llvm.org/D114080

**Comment by Fznamznon** [comment]
Moved to LLORG. Will close this PR in case this is categorically not accepted there.
